src/function/aggregate/nested/list.cpp

```diff
@@ -5,49 +5,791 @@
+	ListVector::Reserve(result, offset);
+
+	// recurse into the linked list of child values
+	D_ASSERT(get_segment_data_function.child_functions.size() == 1);
+	BuildListVector(get_segment_data_function.child_functions[0], &linked_child_list, child_vector, starting_offset);
+}
+
+static void GetDataFromStructSegment(GetSegmentDataFunction &get_segment_data_function, ListSegment *segment,
+                                     Vector &result, idx_t &total_count) {
+
+	auto &aggr_vector_validity = FlatVector::Validity(result);
+
+	// set NULLs
+	auto null_mask = GetNullMask(segment);
+	for (idx_t i = 0; i < segment->count; i++) {
+		if (null_mask[i]) {
+			aggr_vector_validity.SetInvalid(total_count + i);
+		}
+	}
+
+	auto &children = StructVector::GetEntries(result);
+
+	// recurse into the child segments of each child of the struct
+	D_ASSERT(children.size() == get_segment_data_function.child_functions.size());
+	auto struct_children = GetStructData(segment);
+	for (idx_t child_count = 0; child_count < children.size(); child_count++) {
+		auto struct_children_segment = Load<ListSegment *>((data_ptr_t)(struct_children + child_count));
+		auto &child_function = get_segment_data_function.child_functions[child_count];
+		child_function.segment_function(child_function, struct_children_segment, *children[child_count], total_count);
+	}
+}
+
+static void BuildListVector(GetSegmentDataFunction &get_segment_data_function, LinkedList *linked_list, Vector &result,
+                            idx_t &initial_total_count) {
+
+	idx_t total_count = initial_total_count;
+	while (linked_list->first_segment) {
+		auto segment = linked_list->first_segment;
+		get_segment_data_function.segment_function(get_segment_data_function, segment, result, total_count);
+
+		total_count += segment->count;
+		linked_list->first_segment = segment->next;
+	}
+
+	linked_list->last_segment = nullptr;
+}
+
+static void InitializeValidities(Vector &vector, idx_t &capacity) {
+
+	auto &validity_mask = FlatVector::Validity(vector);
+	validity_mask.Initialize(capacity);
+
+	if (vector.GetType().id() == LogicalTypeId::LIST) {
+		auto &child_vector = ListVector::GetEntry(vector);
+		InitializeValidities(child_vector, capacity);
+	} else if (vector.GetType().id() == LogicalTypeId::STRUCT || vector.GetType().id() == LogicalTypeId::MAP) {
+		auto &children = StructVector::GetEntries(vector);
+		for (auto &child : children) {
+			InitializeValidities(*child, capacity);
+		}
+	}
+}
+
+static void RecursiveFlatten(Vector &vector, idx_t &count) {
+
+	if (vector.GetVectorType() != VectorType::FLAT_VECTOR) {
+		vector.Flatten(count);
+	}
+
+	if (vector.GetType().id() == LogicalTypeId::LIST) {
+		auto &child_vector = ListVector::GetEntry(vector);
+		auto child_vector_count = ListVector::GetListSize(vector);
+		RecursiveFlatten(child_vector, child_vector_count);
+
+	} else if (vector.GetType().id() == LogicalTypeId::STRUCT || vector.GetType().id() == LogicalTypeId::MAP) {
```

Comment:
Can also compare the physical type here: physical type of logical struct/map is both physical struct